### ISSUE TITLE
Crypto: Is it safe to pool X509Chain objects without resetting the X509ChainPolicy?

### ISSUE THREAD

**Issue by hiteshmadan** [dotnet/corefx#39927]
I’m building a high-throughput message broker in .net core (2.1 today, upgrading to 3.0-preview7 soon), and need to do client certificate based authentication with a custom cert chain (that I set up on the chain.ChainPolicy.ExtraStore).

To reduce GC pressure, I tend to pool all the byte arrays / batch message arrays / etc.

Is it a good idea to also pool X509Chain objects? And not do a chain.Policy.Reset call on every rent/return?

It’s the X509ChainPolicy.ExtraStore allocation and 5x X509ChainPolicy.ExtraStore.Add(customCert) calls that lead to array resizing, that I want to avoid.

-	When the chain is returned to the pool, I dispose off all the chainElement.Certificate objects and do a chain.Reset() to dispose off the _pal object.
-	I am explicitly resetting chain.ChainPolicy.VerificationTime to DateTime.Now after every pool.Rent().

In other words - Does the X509ChainPolicy ever end up holding on to any valid / invalid state from an old call to chain.Build()?

I couldn’t find anything in the source code that says the Policy leads to any cross-talk between multiple chain.Build() calls using the same Policy object, but still wanted someone from corefx to weigh in and give an all clear.

Thanks!

cc @bartonjs @stephentoub

**Comment by bartonjs** [comment]
`X509ChainPolicy` is just an options input type in the end, nothing is written back to it during chain walking.  It should be fine to keep it long term if (as you already acknowledged) you are explicitly setting VerificationTime before each call to X509Chain.Build.

**Comment by hiteshmadan** [comment]
ok. thanks!
